[FIX] hw_drivers: fallback old display detection method

loouis-t opened the pull request:
In this PR: [https://github.com/odoo/odoo/pull/169633](https://github.com/odoo/odoo/pull/169633)
We updated the method to detect displays to use a safer and better maintained lib. 
For older images (<24.08), that don't have this lib, we chose to create a display anyway as a fallback in case people had one.

We now use the old method as a fallback of the new one to avoid creating a display and starting a browser instance if no screen is plugged in.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/176962.png)](https://mergebot.odoo.com/odoo/odoo/pull/176962)

Ysoroko commented:
tested on image 24_01, works and detects displays as expected
to test on the new image

Ysoroko commented:
Also works with the new image

loouis-t commented:
@odoo/rd-security could you review this please, we added back the old method to detect displays on the IoT boxes.

## Context

In the IoT model, we list devices connected to the IoT Box. Those devices include displays.
To detect them, we used `vcgencmd` lib, which was not compatible with RPI 5.

### Solution we implemented

We added `xrandr` to the new raspios image to be able to detect displays with the python package `screeninfo`.  
However, as old images don't have neither `xrandr` nor `screeninfo`, we caught exceptions and sent a "fake" (or not) display to Odoo db.

### Solution proposed here

We want to fix this by adding back the old method as a fallback to the new one.  
This way we are fully compatible with older linux images.

## Security issue

- `except subprocess.CalledProcessError:`: `vcgencmd` package does not have builtin exceptions, so we need to catch a `subprocess` exception.

Ysoroko commented on addons/hw_drivers/iot_handlers/interfaces/DisplayInterface_L.py:
```diff
@@ -22,28 +26,41 @@ def get_devices(self):
-                    'x_screen': str(x_screen),
-                }
+                display_devices[monitor.name] = self._add_device(monitor.name, x_screen)
+            return display_devices
```
Shouldn't we return a dummy display in case no displays are found here?
```suggestion
            return display_devices if display_devices else dummy_display
```

Ysoroko commented on addons/hw_drivers/iot_handlers/interfaces/DisplayInterface_L.py:
```diff
@@ -22,28 +26,41 @@ def get_devices(self):
+                for x_screen, port in enumerate(hdmi_ports):
+                    if Vcgencmd().display_power_state(hdmi_ports.get(port)) == 'on':
+                        display_devices[port] = self._add_device(port, x_screen)
+                return display_devices
```
Shouldn't we return a dummy display in case no displays are found here?
```suggestion
            return display_devices if display_devices else dummy_display
```

Ysoroko commented on addons/hw_drivers/iot_handlers/interfaces/DisplayInterface_L.py:
```diff
@@ -22,28 +26,41 @@ def get_devices(self):
-            display_devices['distant_display'] = {
-                'name': "Distant Display",
-            }
+            display_devices['distant_display'] = self._add_device('distant_display', 0, distant=True)
```
Can be simplified with a `dummy_display` variable
```suggestion
            return dummy_display
```

Ysoroko commented on addons/hw_drivers/iot_handlers/interfaces/DisplayInterface_L.py:
```diff
@@ -22,28 +26,41 @@ def get_devices(self):
         return display_devices
+
+    @classmethod
+    def _add_device(cls, display_identifier, x_screen, distant=False):
+        """Creates a display_device dict.
+
+        :param display_identifier: the identifier of the display
+        :param x_screen: the x screen number
+        :param distant: if the display is distant (not a real display)
+        :return: the display device dict
+        """
+        if distant:
+            return {'name': display_identifier.replace('_', ' ').capitalize()}
+
```
Can be simplified with a `dummy_display` variable
```suggestion
    def _add_device(cls, display_identifier, x_screen):
        """Creates a display_device dict.

        :param display_identifier: the identifier of the display
        :param x_screen: the x screen number
        :param distant: if the display is distant (not a real display)
        :return: the display device dict
        """
```

qle-odoo reviewed (approved):
@robodoo r+

robodoo commented:
@loouis-t @qle-odoo unable to stage: merge conflict

loouis-t commented:
@robodoo retry

robodoo commented:
@loouis-t @qle-odoo unable to stage: merge conflict

qle-odoo commented:
@loouis-t can you rebase after a fetch

qle-odoo commented:
@robodoo retry

robodoo commented:
@qle-odoo retry makes no sense when the PR is not in error.

qle-odoo commented:
@robodoo r+
